File: modules/alarm-native/ios/AlarmNativeModule.swift
import ExpoModulesCore

/// Mirrors the @objc static API of AlarmKitBridge (app target). The bridge
/// class conforms to this protocol so the pod can call it without a
/// compile-time dependency on the app target.
@objc public protocol AlarmKitBridgeProtocol {
  static func isSupported() -> Bool
  static func requestAuthorization(_ completion: @escaping (Bool) -> Void)
  static func schedule(_ id: String, timestamp: Double, label: String, stopText: String, missionText: String, completion: @escaping (String?) -> Void)
  static func cancel(_ id: String)
  static func cancelAll()
  static func stop(_ id: String)
  static func takeFiredAlarmId() -> String?
}

/// iOS alarms via AlarmKit (iOS 26+). The AlarmKit code itself lives in the app
/// target (plugins/ios/AlarmKitBridge.swift, added by plugins/withAlarmKit.js)
/// because App Intents must be compiled into the app binary.
public class AlarmNativeModule: Module {
  private var observer: NSObjectProtocol?

  private var bridge: AlarmKitBridgeProtocol.Type? {
    NSClassFromString("AlarmKitBridge") as? AlarmKitBridgeProtocol.Type
  }

  public func definition() -> ModuleDefinition {
    Name("AlarmNative")
    Events("onAlarmLaunch")

    OnCreate {
      self.observer = NotificationCenter.default.addObserver(forName: Notification.Name("missionalarm.fired"), object: nil, queue: .main) { [weak self] note in
        if let id = note.userInfo?["alarmId"] as? String {
          self?.sendEvent("onAlarmLaunch", ["alarmId": id])
        }
      }
    }

    OnDestroy {
      if let o = self.observer { NotificationCenter.default.removeObserver(o) }
    }

    Function("isSupported") { self.bridge?.isSupported() ?? false }

    AsyncFunction("requestAuthorization") { (promise: Promise) in
      guard let bridge = self.bridge else { return promise.resolve(false) }
      bridge.requestAuthorization { granted in promise.resolve(granted) }
    }

    Function("schedule") { (id: String, triggerAt: Double, label: String, stopText: String, missionText: String) in
      self.bridge?.schedule(id, timestamp: triggerAt, label: label, stopText: stopText, missionText: missionText) { error in
        if let error { NSLog("AlarmKit schedule failed: %@", error) }
      }
    }

    Function("cancel") { (id: String) in self.bridge?.cancel(id) }
    Function("cancelAll") { self.bridge?.cancelAll() }
    Function("stop") { (id: String) in self.bridge?.stop(id) }
    Function("takeFiredAlarmId") { () -> String? in self.bridge?.takeFiredAlarmId() }

    Function("dismissNotification") {}
    Function("canScheduleExact") { true }
    Function("openExactAlarmSettings") {}
    Function("showOverLockScreen") { (_: Bool) in }
  }
}

File: modules/alarm-native/src/AlarmNativeModule.ts
import { NativeModule, requireOptionalNativeModule } from 'expo';

export type AlarmNativeEvents = {
  onAlarmLaunch: (e: { alarmId: string }) => void;
};

declare class AlarmNativeModule extends NativeModule<AlarmNativeEvents> {
  isSupported(): boolean;
  schedule(id: string, triggerAt: number, label: string, stopText: string, missionText: string): void;
  cancel(id: string): void;
  cancelAll(): void;
  takeFiredAlarmId(): string | null;
  stop(id: string): void;
  requestAuthorization(): Promise<boolean>;
  dismissNotification(): void;
  canScheduleExact(): boolean;
  openExactAlarmSettings(): void;
  showOverLockScreen(enabled: boolean): void;
}

const native = requireOptionalNativeModule<AlarmNativeModule>('AlarmNative');

/** Safe wrapper: no-ops when the native module is absent (Expo Go / web). */
export const AlarmNative = {
  isSupported: () => !!native && native.isSupported(),
  /** stopText / missionText label the buttons of the OS alarm alert (iOS AlarmKit). */
  schedule: (id: string, triggerAt: number, label: string, stopText: string, missionText: string) =>
    native?.schedule(id, triggerAt, label, stopText, missionText),
  cancel: (id: string) => native?.cancel(id),
  cancelAll: () => native?.cancelAll(),
  takeFiredAlarmId: () => native?.takeFiredAlarmId() ?? null,
  dismissNotification: () => native?.dismissNotification(),
  /** Stop the OS-level alert for this alarm (JS ring screen has taken over). */
  stop: (id: string) => native?.stop(id),
  requestAuthorization: () => native?.requestAuthorization() ?? Promise.resolve(false),
  canScheduleExact: () => native?.canScheduleExact() ?? true,
  openExactAlarmSettings: () => native?.openExactAlarmSettings(),
  showOverLockScreen: (enabled: boolean) => native?.showOverLockScreen(enabled),
  addLaunchListener: (fn: (e: { alarmId: string }) => void) => native?.addListener('onAlarmLaunch', fn) ?? { remove() {} },
};

export default AlarmNative;
